src/emqx_router.erl

```diff
@@ -266,28 +266,34 @@ maybe_trans(Fun, Args) ->
                   end, [])
     end.
 
+%% The created fun only terminates with explicit exception
+-dialyzer({nowarn_function, [trans/2]}).
```

Comment:
like dialyzer, I have a gut feeling it is not a good idea to let the process exits abnormally. 😄 

I haven't got a chance to look into the VM yet, maybe  @gorillainduction has some insights? 

If there is not much overhead or side effects, I think we can go for this approach.

Reply:
`{shutdown, Reason}` is not a very common exit reason.

Reply:
The main benefit of this approach is that the code now uses "the ref trick" for receiving the result (i.e., what zmstone is refering to as "receive-mark optimization" below). It is a bit ugly, but the benefits can be huge.

I checked the beam assembly code, and the new code is indeed setting a receive mark, whereas the old one didn't

```
                 {function,trans,2,90,
                            [{label,89},
                             {line,[{location,"emqx_router.erl",273}]},
                             {func_info,{atom,emqx_router},{atom,trans},2},
                             {label,90},
                             {'%',{var_info,{x,0},[{type,{t_fun,any,any}}]}},
                             {'%',{var_info,{x,1},
                                            [{type,{t_list,{t_tuple,3,true,#{1 => {t_atom,[route]}}},
                                                           nil}}]}},
                             {allocate_zero,2,2},
                             {make_fun2,{f,110},0,0,2},
                             {'%',{var_info,{x,0},[{fun_type,none}]}},
                             {recv_mark,{f,91}},                       <------ HERE
                             {line,[{location,"emqx_router.erl",275}]},
                    ...
```